src/SpaceStation.cpp

```diff
@@ -383,72 +409,72 @@ bool SpaceStation::GetDockingClearance(Ship *s)
+	}
+	if (IsPortLocked(bay)) {
+		return DoShipDamage(s, flags, relVel);
+	}
+	if (m_shipDocking[bay].stage == DockStage::NONE) {
+		return DoShipDamage(s, flags, relVel); // also no clearance?
+	}
+
+	if (IsGroundStation() || touchOrbitalPad) {
+		// must be oriented sensibly and have wheels down
+		matrix4x4d bayTrans = GetBayTransform(bay);
+
+		vector3d dockingNormal = bayTrans.Up();
+		const double dot = s->GetOrient().VectorY().Dot(dockingNormal);
+		if ((dot < 0.99) || (s->GetWheelState() < 1.0))
+		{
+			return DoShipDamage(s, flags, relVel); // <0.99 harsh?
 		}
-		if (IsPortLocked(port)) {
+		// check speed
+		if (s->GetVelocity().Length() > MAX_LANDING_SPEED)
+		{
 			return DoShipDamage(s, flags, relVel);
 		}
-		if (m_shipDocking[port].stage != 1) return DoShipDamage(s, flags, relVel); // already docking?
-
-		SpaceStationType::positionOrient_t dport;
-
-		if (IsGroundStation()) {
-			// must be oriented sensibly and have wheels down
-			PiVerify(m_type->GetDockAnimPositionOrient(port, m_type->NumDockingStages(), 1.0f, vector3d(0.0), dport, s));
-			vector3d dockingNormal = GetOrient() * dport.yaxis;
-			const double dot = s->GetOrient().VectorY().Dot(dockingNormal);
-			if ((dot < 0.99) || (s->GetWheelState() < 1.0)) return DoShipDamage(s, flags, relVel); // <0.99 harsh?
-			// check speed
-			if (s->GetVelocity().Length() > MAX_LANDING_SPEED) return DoShipDamage(s, flags, relVel);
-			// check if you're near your pad
-			float dist = (s->GetPosition() - GetPosition() - GetOrient() * dport.pos).LengthSqr();
-			// docking allowed only if inside a circle 70% greater than pad itself (*1.7)
-			float maxDist = static_cast<float>(m_type->FindPortByBay(port)->maxShipSize / 2) * 1.7;
-			if (dist > (maxDist * maxDist)) return DoShipDamage(s, flags, relVel);
+		// check if you're near your pad
+		float dist = (s->GetPosition() - bayTrans.GetTranslate()).LengthSqr();
+		// docking allowed only if inside a circle 70% greater than pad itself (*1.7)
+		float maxDist = static_cast<float>(m_type->FindPortByBay(bay)->maxShipSize / 2) * 1.7;
+		if (dist > (maxDist * maxDist))
+		{
+			return DoShipDamage(s, flags, relVel);
 		}
+	}
 
-		// why stage 2? Because stage 1 is permission to dock
-		// granted, stage 2 is start of docking animation.
-		PiVerify(m_type->GetDockAnimPositionOrient(port, 2, 0.0, vector3d(0.0), dport, s));
+	// docking is in progress
+	if (s->GetFlightState() == Ship::DOCKING) return false;
+
+	// launch docking
+	shipDocking_t &sd = m_shipDocking[bay];
+	sd.ship = s;
+	sd.stagePos = 0;
+	sd.fromPos = (s->GetPosition() - GetPosition()) * GetOrient(); // station space
+	sd.fromRot = Quaterniond::FromMatrix3x3(GetOrient().Transpose() * s->GetOrient());
+	s->SetFlightState(Ship::DOCKING);
+	s->SetVelocity(vector3d(0.0));
+	s->SetAngVelocity(vector3d(0.0));
+	s->ClearThrusterState();
+	SwitchToStage(bay, (touchOrbitalPad || m_type->NumUndockStages() == 0) ? DockStage::TOUCHDOWN : DockStage::DOCK_ANIMATION_1);
```

Comment:
Is this block of code worth separating to its own function? If not, I'd recommend adding a slightly more verbose preceeding comment as it's very dense and not fully clear *what* is being done here.

Reply:
Since it is not reused anywhere, I did not separate it into a function and just added comments.